### ISSUE TITLE
Does it make sense to have an `Arc<>` for each service in `Context` if `Context` itself is an `Arc<>`?

### ISSUE THREAD

**Issue by frederikhors** [bkonkle/rust-example-caster-api#12]
First of all: THANKS FOR YOUR PROJECT! I'm learning a lot from it! THANKS!!!

A small question (I'm very new to Rust):

Given the code:

```rust
pub struct Context {
    pub config: &'static Config,
    pub db: Arc<DatabaseConnection>,
    pub oso: Oso,
    pub users: Arc<dyn UsersServiceTrait>,
    pub profiles: Arc<dyn ProfilesService>,
    pub role_grants: Arc<dyn RoleGrantsService>,
    pub shows: Arc<dyn ShowsService>,
    pub episodes: Arc<dyn EpisodesService>,
}

impl Context {
    pub async fn init(config: &'static Config) -> Result<Self> {
        let db = Arc::new(sea_orm::Database::connect(&config.database.url).await?);

        //...

        Ok(Self {
            config,
            users: Arc::new(UsersService::new(&db)),
            profiles: Arc::new(DefaultProfilesService::new(&db)),
            role_grants: Arc::new(DefaultRoleGrantsService::new(&db)),
            shows: Arc::new(DefaultShowsService::new(&db)),
            episodes: Arc::new(DefaultEpisodesService::new(&db)),
            oso,
            db,
        })
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    //...

    let context = Arc::new(Context::init(config).await?);

    let (addr, server) = run(context).await?;

    //...
}
```

does it make sense to have an `Arc<>` for each service in `Context` if `Context` itself is an `Arc<>`?

Thanks endlessly.

**Comment by bkonkle** [comment]
Hi! I apologize for the slow response here. I just saw this issue! I'll take a look a bit later this week, and see if I can find a way to pull the services out of the nested Arcs. Thank you for pointing it out!

**Comment by bkonkle** [comment]
👋 Hi! I tackled this issue in my first Live Coding session! Here's the recording on YouTube: https://www.youtube.com/watch?v=R8ZRU3tpqDQ

TLDR: In order to provide easy test dependency injection, I'm pulling individual dependencies out of the `Context` struct and passing them in to things like the GraphQL Dataloader and the Schema context that async-graphql provides. Because of this, I need the shared ownership that an Arc provides. It also needs to be thread-safe because the Warp library spawns a different thread for each HTTP request, so that rules out the more flexible Rc type, which handles shared ownership without thread safety.

**Comment by frederikhors** [comment]
I saw your video. Thank you very much!
